### ISSUE TITLE
Improve general performance around short URLs for instances with a large number

### ISSUE THREAD

**Issue by marijnvandevoorde** [shlinkio/shlink#2112]
### Shlink version

4.1.0

### PHP version

8.x

### How do you serve Shlink

Docker image

### Database engine

MySQL

### Database version

8.x

### Current behavior

The recently introduced DeleteExpiredShortUrlsCommand to clean up expired links is great, as we foresee our database to grow in the millions over time. However, as there is no index on the expiry date column, this command will become increasingly heavy as it has to go over every single record.

### Expected behavior

An index on the expiry date column will drastically increase the speed of the DeleteExpiredShortUrlsCommand

### Minimum steps to reproduce

Fill up the database with a couple of thousand/million links.
Run the DeleteExpiredShortUrlsCommand
Now add the index and compare the performance of the DeleteExpiredShortUrlsCommand

**Comment by acelaya** [comment]
Thanks!

> An index on the expiry date column will drastically increase the speed

Yep, I think you are right there. I'll do some testing.

**Comment by marijnvandevoorde** [comment]
I took the liberty of adding it, hope I didn't overlook any guidelines you have while doing so. If not, it's not the biggest change and the PR can maybe provide you some ready-to-copy-paste material ;-)

**Comment by acelaya** [comment]
Yep, I saw it, thanks a lot 🙂. I was definitely planning to use that migration to test this.

**Comment by marijnvandevoorde** [comment]
I was super surprised by your message at first, but I think I have an explanation:
Since all links are expired, the command has to delete all of them, so it has to go over all links, regardless of the presence of an index. 

The real test would be to have 1M links but only 1k expired, and then run it? It should be almost instand with an index but will probably take 20 seconds without the index?

**Comment by acelaya** [comment]
Hmm, yeah, that actually makes sense. I'll try that and come back with the results.

I also want to experiment a bit with short urls which are "expired" due to having reached the max amount of visits. In that case, the query requires a sub query, so it'll require a different optimization approach.

**Comment by acelaya** [comment]
> I was super surprised by your message at first, but I think I have an explanation: Since all links are expired, the command has to delete all of them, so it has to go over all links, regardless of the presence of an index.
> 
> The real test would be to have 1M links but only 1k expired, and then run it? It should be almost instand with an index but will probably take 20 seconds without the index?

Well, the execution times are again very similar 😅 

To be honest, with just 1k expired URLs and the rest having `valid_since` in the future, the query takes ~30ms with the index, and ~150ms without it, so it somewhat has an effect, but it is not terribly bad without it.

And then, the more expired short URLs there are, the less the index makes a difference. I then tried with 100k expired short URLs from the 1M total URLs, and it took ~2s in both cases.

Again, I just tested with Postgres, and it could be this difference is bigger with MySQL, specially with >=8.0 versions.

I feel tempted to merge the PR anyway, with a couple changes, as it definitely doesn't make things worst, but I feel like I should test other database engines first.

In the meantime, feel free to do some tests yourself, as I might be making some mistakes.

**Comment by marijnvandevoorde** [comment]
We were planning to do some performance testing so we can do that as well. To be honest, the PR also comes from "this should have an impact" rather than from experiencing issues. 

Unrelated, we also noticed there's no index on the long urls. Shouldn't that be there for people who enable the "findIfExists" feature? I can imagine that becoming slow when the number of urls starts to rise. Again, just what "I would expect", haven't done any testing. We just anticipate serving several millions of links over time and we're checking if there are any reasons NOT to use shlink :-)

**Comment by acelaya** [comment]
I'm going to re-purpose this issue with a more general context around short URLs. I'll list the different improvements as they have been applied.
